### PR TITLE
Add new geometries and tests

### PR DESCRIPTION
**Pull request by apcraig**
- Add wall kmt_type
    - Add blockep4, uniformp5, medblocke, blocke to ice_data_type
    - Add ability to check if C/CD fields are on restart file and skip
    - Add new tests to gridsys_suite
    - Update documentation with new namelist

### CODE REVIEW

File: configuration/scripts/options/set_nml.box2000
Comment (eclare108213) [inline configuration/scripts/options/set_nml.box2000]: why is this file named box2000?

File: cicecore/cicedynB/infrastructure/ice_restart_driver.F90
Comment (eclare108213) [inline cicecore/cicedynB/infrastructure/ice_restart_driver.F90:362]: Nice.  Do restarts now work for C- and CD-grids?

**Comment by apcraig** [comment]
Some test suite results.  kmtislands is completing where it wasn't before.  I think the recent fixes to the halo update and otherwise helped quite a bit with those.  https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#dc29012a3717628e65a0c11a33ea84a394a4f9e9

File: cicecore/cicedynB/infrastructure/ice_restart_driver.F90
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_restart_driver.F90:362]: On restart, the C/CD restart tests are aborting.  @dabail10 suggested C/CD restarts are working, so we need to just check that.  The fix here allows the C/CD cases to start from B restarts which is a (temporary?) requirement to get some of the test cases to run.

File: configuration/scripts/options/set_nml.box2000
Comment (apcraig) [reply configuration/scripts/options/set_nml.box2000]: Hunke 2000?  Happy to name it something else.  I assume that's where box2001 came from, Hunke 2001?

**Comment by apcraig** [comment]
@dabail10, if you get a chance, maybe you can check some of the restart tests in the gridsys test suite.  There are just a few,

```
restart        gx3     4x2        debug,diag1
restart        gx3     4x2        debug,diag1,gridcd
restart        gx3     4x2        debug,diag1,gridc
```

The B grid works fine, the C and CD tests abort in my testing.  I don't think we need to hold up this PR, but it would be good for the restart tests to pass if they are working.  Thanks.

File: cicecore/cicedynB/infrastructure/ice_restart_driver.F90
Comment (dabail10) [reply cicecore/cicedynB/infrastructure/ice_restart_driver.F90:362]: So, these only work for ice_ic = 'default'. I had that in the PR. I had not added the capability for reading the B-grid initial files. This is one piece, which is great, but I was thinking that for a CD or C test reading the B-grid restart that we would initialize the C/CD velocities from the B-grid.

Comment (dabail10) [reply cicecore/cicedynB/infrastructure/ice_restart_driver.F90:362]: Thanks for adding this.

**Review by dabail10 — APPROVED** [review]

File: cicecore/cicedynB/infrastructure/ice_restart_driver.F90
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_restart_driver.F90:362]: The gridsys_suite starts from B restarts.  I agree we could do more to initialize from B, but the C/CD smoke tests from B restarts run fine.  I don't understand why ice_ic='default' plays a role in the restart test.  The C/CD restart test starts from a B restart fine and writes a restart after 5 days (or so).  When the 2nd run of the restart test starts up, it definitely reads the C/CD fields from the new restart file then aborts in the shortwave code somewhere.  I think with these fixes, the ice_ic='default' problem should be fixed.  What restart tests were you running?  Can you review the restart tests in the gridsys_suite and see how they differ from your test when you have a chance.  Thanks.

**Comment by dabail10** [comment]
ice_ic = 'default' is the internal initialization like we do for the box tests. If you have ice_ic = 'filename' as we do for the gx3 this reads the B-grid initial files. It is only the initial step that requires this. After this the restarts all work fine. So, we need your hack to read a B initial file, or even better we will create C / CD grid initial files eventually.

File: configuration/scripts/options/set_nml.box2000
Comment (eclare108213) [reply configuration/scripts/options/set_nml.box2000]: Yes, box2001 is the test case used in Hunke (2001).  I only had one paper in 2000, and it was about penguin moulting habitat!!!  So let's come up with some other name.  Is this the 1D test in appendix A?  Maybe call it box1Dwall and put a comment in the code that it's appendix A in Hunke (2001), and note any differences, e.g. there's only ice in half the domain.

Comment (apcraig) [reply configuration/scripts/options/set_nml.box2000]: OK, I'll update the name. I see the paper is Hunke 2001, I thought it was 2000.  Sorry about that.

**Comment by apcraig** [comment]
@dabail10.  Again, the problem with restarts is NOT starting from the B restarts.  The "query_field" addresses that.  The B restarts are read fine and the smoke tests pass.  In that case, the velocity fields are set to zero at the start (just as they would be in any arbitrary initialization) and that is reasonable.  The problem with the restarts is that when the C/CD restart file is read as part of the second run of the restart test, the code aborts in the physics somewhere.  

I understand restarts should be working, but in my test, they are not.  Could you have a look and see whether you can understand why restarts work for you but not for the tests we have setup in the gridsys_suite.  Thanks.

File: configuration/scripts/options/set_nml.box2000
Comment (eclare108213) [reply configuration/scripts/options/set_nml.box2000]: Now I'm a little confused.  We've been using box2001 as the standard box test case from that paper, with land all around, cyclonic (or maybe anticyclonic) ocean currents, time dependent atmospheric forcing, etc.  I was thinking we should reserve that name for that specific test case and use a different name for the 1D test in the appendix.  Is there (or will there be) also a set_nml.box2001 for the other test?

**Comment by dabail10** [comment]
Ah. Ok. I will check that.

File: configuration/scripts/options/set_nml.box2000
Comment (apcraig) [reply configuration/scripts/options/set_nml.box2000]: box2001 has and will remain unchanged.  it's the closed box test as it's always been.  I will rename box2000 to boxwallp5.

**Review by phil-blain — COMMENTED** [review]
I had a quick look and this looks good.

**Comment by dabail10** [comment]
This is weird. My restart test here:

/glade/work/dbailey/CICE_Cgrid/cheyenne_intel_restart_gx3_4x1_debug_gridcd.gx3rest

ran fine. So, maybe the recent changes broke this? Looks like it is dying here:

      hol = vonkar * gravit * zlvl &
               * (tstar/thva &
               + qstar/(c1/zvir+Qa)) &
               / ustar**2

in the division by ustar. ustar is computed from rdn*vmag which should not be zero ... there is also division by thva, and zvir. Did something change in ice_forcing where these are coming out as zero?

Dave

**Comment by dabail10** [comment]
My test is cheyenne intel and yours is cheyenne_gnu. Trying that now.

**Comment by apcraig** [comment]
I think we should merge this unless other changes are needed.  Any fixes to get exact restart working should come in as a separate PR anyway.  Anyone disagree?

**Comment by dabail10** [comment]
Just an update here. I am finding some divide by zero issues in icepack. I am puzzled why this is only a problem when doing a restart with CD/C grids. I assume there is some mask issue where icepack is being called for a non-ice point?

**Comment by dabail10** [comment]
This is weird. The Tair is being initialized weird. Looks like a block problem. Note this is only in the restart initial file.

Screen Shot 2022-02-11 at 10 48 16 AM

**Comment by apcraig** [comment]
I'm going to merge and we can deal with the restart issues separately.  Thanks for all your effort @dabail10.  I will also try to look at what's happening.  Given only the first block is written, is there a "master task" logical somewhere where it shouldn't be?  Just speculation.  That tair plot is in the B restart or the CD restart or after reading the CD restart?

**Comment by dabail10** [comment]
I will start a new issue. The B-initial file looks fine and the first initial file from C/CD looks fine. So, there is something weird with the C/CD restart with initializing the atmosphere data.

**Pull request merged**
